### PR TITLE
Jobs

### PR DESCRIPTION
**Pull request by GriffinCodes**
Closes ProjectEdenGG/Issues#483

- Replaces yucky MySQL based Task/TaskService
- Jobs are abstract classes that are stored in Mongo
    - Class name kept in Mongo to tell the object mapper which class to deserialize as
        - Refactoring/renaming classes is dangerous! If it can't find the class it will error
            - This is partly why I chose to leave all jobs in the same folder, instead of near their respective codebase. Less prone to random refactoring
- Two types of jobs:
    - Automatically scheduled jobs (`@Schedule`)
        - Checks every minute if the next task is scheduled, and schedules it if not
            - minute == unix cron precision
    - Manually scheduled jobs (`AbstractJob#schedule`)
        - One second precision
- Handles errors & interruptions as gracefully as possible
- Uses CompleteableFutures to ensure async/delayed tasks are completed properly, or marked as interrupted
    - Annotate such jobs with `@RetryIfInterrupted` to rerun task (only if its safe to do so!)
    
### TODO
 - I want to make the command a lot more useful, but that can be done after merge

### CODE REVIEW

**Review by WakkaFlocka239 — APPROVED** [review]

File: src/main/java/gg/projecteden/nexus/models/scheduledjobs/ScheduledJobs.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+@Entity(value = "scheduled_jobs", noClassnameStored = true)
+@NoArgsConstructor
+@AllArgsConstructor
+@RequiredArgsConstructor
+@Converters({UUIDConverter.class, JobConverter.class, LocalDateTimeConverter.class})
+public class ScheduledJobs implements PlayerOwnedObject {
+	@Id
+	@NonNull
+	private UUID uuid;
+	private Map<JobStatus, Set<AbstractJob>> jobs = new HashMap<>();
```
Comment (lexikiq) [inline src/main/java/gg/projecteden/nexus/models/scheduledjobs/ScheduledJobs.java:38]: worried this object may grow too large for the database, but given that the only time this has happened was with cyn's transaction history and that we don't have an alternative system rn, it's probably fine

File: src/main/java/gg/projecteden/nexus/features/commands/staff/admin/JobsCommand.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+				service.save(jobs);
+			}
+		} catch (Exception ex) {
+			ex.printStackTrace();
+		}
+	}
+
+	private static void processor() {
+		Tasks.repeat(0, Time.SECOND, () -> {
```
Comment (lexikiq) [inline src/main/java/gg/projecteden/nexus/features/commands/staff/admin/JobsCommand.java:66]: seems like previous implementations of tasks were async, but this is sync, is that intentional?

**Review by lexikiq — APPROVED** [review]
> Refactoring/renaming classes is dangerous! If it can't find the class it will error 
a warning about this in the javadocs on the AbstractJob class would be useful as new devs or devs not very experienced with the framework may forget this

overall looks great

File: src/main/java/gg/projecteden/nexus/models/scheduledjobs/common/AbstractJob.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	private ScheduledJobs jobs() {
+		return new ScheduledJobsService().get0();
+	}
+
+	public void setStatus(JobStatus status) {
+		jobs().get(this.status).remove(this);
+		this.status = status;
+		jobs().get(this.status).add(this);
```
[the remark this answers is not in the document]
Comment (GriffinCodes) [reply src/main/java/gg/projecteden/nexus/models/scheduledjobs/common/AbstractJob.java:49]: No, thats not the job of this method. Things that call this method need to save it when they are done

File: src/main/java/gg/projecteden/nexus/models/scheduledjobs/ScheduledJobs.java (hunk 1 above)
Comment (GriffinCodes) [reply src/main/java/gg/projecteden/nexus/models/scheduledjobs/ScheduledJobs.java:38]: https://github.com/ProjectEdenGG/Nexus/pull/20/files/2a10fafbd9fec602746bccfc3d45255b586c2e63#diff-ace30cdf0a5982ea89aec7db8032e23221c439cbde970fdbd7435add0581e346R66-R69

I will manually keep an eye on the errored/interrupted jobs, and will probably add them to the janitor at some point when i feel more confident about the system

File: src/main/java/gg/projecteden/nexus/features/commands/staff/admin/JobsCommand.java (hunk 2 above)
Comment (GriffinCodes) [reply src/main/java/gg/projecteden/nexus/features/commands/staff/admin/JobsCommand.java:66]: I didn't feel the need to make any of the current tasks asynchronous, they are small enough, or in one case spawn their own async task. Will keep an eye on it tho

**Pull request merged**